std/assembly/string.ts

```diff
@@ -393,7 +394,127 @@ export class String {
     return result;
   }
 
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replace(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var index = this.indexOf(search);
+    if (~index) {
+      let rlen = replacement.length;
+      len = len - slen;
+      let olen = len + rlen;
+      if (olen) {
+        let result = allocateUnsafe(olen);
+        copyUnsafe(result, 0, this, 0, index);
+        copyUnsafe(result, index, replacement, 0, rlen);
+        copyUnsafe(result, rlen + index, this, index + slen, len - index);
+        return result;
+      }
+    }
+    return this;
+  }
+
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replaceAll(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var rlen = replacement.length;
+    if (!slen) {
+      // Handle special case when we should insert replacement between each char
+      // TODO handle surrogate pairs as well
+      if (!rlen) return this;
+      let result = allocateUnsafe(len + (len + 1) * rlen);
+      copyUnsafe(result, 0, replacement, 0, rlen);
+      let offset = rlen;
+      for (let i = 0; i < len; ++i) {
+        store<u16>(
+          changetype<usize>(result) + (offset++ << 1),
+          load<u16>(changetype<usize>(this) + (i << 1), HEADER_SIZE),
+          HEADER_SIZE
+        );
+        copyUnsafe(result, offset, replacement, 0, rlen);
+        offset += rlen;
+      }
+      return result;
+    }
+    var prev = 0, next = 0;
+    if (slen === rlen) {
+      // Fast path when search and replacement have same length
+      let result = allocateUnsafe(len);
+      copyUnsafe(result, 0, this, 0, len);
+      while (~(next = this.indexOf(search, prev))) {
+        copyUnsafe(result, next, replacement, 0, rlen);
+        prev = next + slen;
+      }
+      return result;
+    }
+    var result = allocateUnsafe(len);
+    var offset = 0, resLen = len;
+    while (~(next = this.indexOf(search, prev))) {
+      if (offset > resLen) {
+        // resize
+        let newLength = resLen << 1;
+        let newResult = allocateUnsafe(newLength);
+        copyUnsafe(newResult, 0, result, 0, offset);
+        freeUnsafe(result);
```

Comment:
Hmm, in a GC scenario this does nothing. Is there maybe a better way than allocating in a loop?

Reply:
I guess that's not so easy because there's no reallocateUnsafe for strings?

Reply:
yes, I implemented `reallocateUnsafe` in the past but it seems this not was good idea, I don't remember)